app/models/manageiq/providers/hawkular/inventory/server_operations.rb

```diff
@@ -0,0 +1,278 @@
+module ManageIQ
+  module Providers
+    module Hawkular
+      module Inventory
+        module ServerOperations
+          extend ActiveSupport::Concern
+
+          class_methods do
+            def group_operation(name, *args)
+              generic_operation("#{name}_middleware_server_group", *args)
+            end
+
+            def domain_operation(name, *args)
+              generic_operation("#{name}_middleware_domain_server", *args)
+            end
+
+            def standalone_operation(name, *args)
+              generic_operation("#{name}_middleware_server", *args)
+            end
+
+            def generic_operation(name, action_name, default_params = {}, default_extra_data = {})
+              define_method(name) do |ems_ref, params = {}, extra_data = {}|
+                run_generic_operation(action_name.to_sym, ems_ref, default_params.merge(params), default_extra_data.merge(extra_data))
+              end
+            end
+
+            def specific_operation(name, action_name, default_params = {})
+              define_method(name) do |ref, params = {}|
+                params[:resourcePath] = ref.to_s
+                run_operation(default_params.merge(params), action_name)
+              end
+            end
+          end
+
+          NotificationArgs = Struct.new(:type, :operation_name, :operation_args, :target_resource, :entity_klass, :detailed_message) do
+            def self.success(*args)
+              new(:mw_op_success, *args)
+            end
+
+            def event_type(entity)
+              attributes = {
+                :entity_type => entity.kind_of?(MiddlewareServer) ? 'MwServer' : 'MwDomain',
+                :operation   => operation_name,
+                :status      => type == :mw_op_success ? 'Success' : 'Failed'
+              }
+
+              '%{entity_type}.%{operation}.%{status}' % attributes
+            end
+
+            def event_message(entity)
+              attributes = {
+                :operation => operation_name,
+                :server    => entity.name,
+                :status    => type == :mw_op_success ? _('succeeded') : _('failed')
+              }
+
+              message = _('%{operation} operation for %{server} %{status}') % attributes
```

Comment:
what is the `% attributes` part does here?

Reply:
Named string interpolation. It works like `printf`, `sprintf` and so on.